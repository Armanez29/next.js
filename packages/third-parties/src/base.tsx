import React from 'react'

export default function Base({
  content,
  height = null,
  width = null,
  children,
}: {
  content?: string
  height?: number | null
  width?: number | null
  children?: React.ReactElement | React.ReactElement[]
}) {
  return (
    <>
      {/* insert script children */}
      {children}
      {/* insert content */}
      {content && (
        <div
          style={{
            height: `${height}px` || 'auto',
            width: `${width}px` || 'auto',
          }}
          dangerouslySetInnerHTML={{ __html: content }}
        />
      )}
    </>
  )
}
