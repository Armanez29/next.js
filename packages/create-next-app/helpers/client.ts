/* eslint-disable import/no-extraneous-dependencies */
import got from 'got'
import { HttpsProxyAgent } from 'hpagent'

function getAgent() {
  if (process.env.https_proxy) {
    return {
      https: new HttpsProxyAgent({
        proxy: process.env.https_proxy,
        rejectUnauthorized: false,
      }),
    }
  }
}

export default got.extend({ agent: getAgent() })
