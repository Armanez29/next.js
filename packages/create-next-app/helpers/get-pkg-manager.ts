export type PackageManager = 'npm' | 'pnpm' | 'yarn'

export function getPkgManager(): PackageManager {
  const userAgent = process.env.npm_config_user_agent

  if (userAgent && userAgent.startsWith('yarn')) {
    return 'yarn'
  }
  
  if (userAgent && userAgent.startsWith('pnpm')) {
    return 'pnpm'
  } 
  
  return 'npm'
}
