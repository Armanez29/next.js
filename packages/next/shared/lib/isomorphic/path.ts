/**
 * This module is for next.js server internal usage of path module.
 * It will use native path module for nodejs runtime.
 * It will use path-browserify polyfill for edge runtime.
 */

import type { PlatformPath } from 'path'

const path: PlatformPath =
  process.env.NEXT_RUNTIME === 'edge'
    ? require('next/dist/compiled/path-browserify')
    : require('path')

export default path

const {
  join,
  resolve,
  normalize,
  isAbsolute,
  relative,
  dirname,
  basename,
  sep,
  delimiter,
  extname,
  parse,
  format,
  toNamespacedPath,
  win32,
  posix,
} = path

export {
  join,
  resolve,
  normalize,
  isAbsolute,
  relative,
  dirname,
  basename,
  sep,
  delimiter,
  extname,
  parse,
  format,
  toNamespacedPath,
  win32,
  posix,
}
