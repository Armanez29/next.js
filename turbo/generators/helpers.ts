import type { PlopTypes } from '@turbo/gen'

export function toFileName(str: string) {
  return str.toLowerCase().replace(/ /g, '-')
}

export function init(plop: PlopTypes.NodePlopAPI): void {
  plop.setHelper('toFileName', toFileName)
}
