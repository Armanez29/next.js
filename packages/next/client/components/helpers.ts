import type { CacheNode } from '../../shared/lib/app-router-context'
import type { FlightRouterState } from '../../server/app-render'
import { fetchServerResponse } from './app-router.client'

export const createOptimisticTree = (
  existingState: any,
  url: URL,
  // /dashboard/integrations ['', 'dashboard', 'integrations', 'page']
  segments: string[],
  existingFlightRouterState: FlightRouterState | null,
  childNodes: CacheNode['childNodes'] | null,
  isFirstSegment: boolean,
  parentFetchingData: boolean,
  initialMap = new Map()
): {
  childNodes: CacheNode['childNodes']
  tree: FlightRouterState
} => {
  const [existingSegment, existingParallelRoutes] =
    existingFlightRouterState || [null, {}]
  const segment = segments[0]
  const isLastSegment = segments.length === 1

  let parallelRoutes: FlightRouterState[1] = {}
  if (existingSegment === segment) {
    parallelRoutes = existingParallelRoutes
  }

  if (childNodes) {
    for (const [key, value] of childNodes) {
      initialMap.set(key, {
        subTreeData: value.subTreeData,
        childNodes: new Map(),
      })
    }
  }

  let startedFetch = false
  if (!isFirstSegment && !parentFetchingData) {
    // Seed the cache with missing item
    if (childNodes && !childNodes.has(segment)) {
      startedFetch = true
      console.log('KICKING OFF DATA FETCH IN PUSH', {
        segment,
        url,
      })

      initialMap.set(segment, {
        data: fetchServerResponse(url, existingState.tree),
        subTreeData: null,
        childNodes: new Map(),
      })
    }
  }

  let childTree
  if (!isLastSegment) {
    const passedChildNodes = isFirstSegment
      ? childNodes
      : childNodes && childNodes.has(segment)
      ? childNodes.get(segment)!.childNodes
      : null
    const childItem = createOptimisticTree(
      existingState,
      url,
      segments.slice(1),
      parallelRoutes ? parallelRoutes.children : null,
      passedChildNodes,
      false,
      parentFetchingData || startedFetch,
      isFirstSegment ? initialMap : initialMap.get(segment)!.childNodes
    )

    childTree = childItem.tree
  }

  const result: FlightRouterState = [
    segment,
    {
      ...parallelRoutes,
      ...(childTree ? { children: childTree } : {}),
    },
  ]

  // Add url into the tree
  if (isFirstSegment) {
    result[2] = url.pathname + url.search
  }

  return {
    childNodes: initialMap,
    tree: result,
  }
}
