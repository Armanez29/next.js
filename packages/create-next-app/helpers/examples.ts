/* eslint-disable import/no-extraneous-dependencies */
import tar from 'tar'
import { Stream, Readable } from 'stream'
import { promisify } from 'util'
import { join } from 'path'
import { tmpdir } from 'os'
import { createWriteStream, promises as fs } from 'fs'

const pipeline = promisify(Stream.pipeline)

export type RepoInfo = {
  username: string
  name: string
  branch: string
  filePath: string
}

export async function isUrlOk(url: string): Promise<boolean> {
  try {
    const res = await fetch(url, { method: 'HEAD' })
    if (res.status !== 200) {
      return false
    }
    return true
  } catch {
    return false
  }
}

export async function getRepoInfo(
  url: URL,
  examplePath?: string
): Promise<RepoInfo | undefined> {
  const [, username, name, t, _branch, ...file] = url.pathname.split('/')
  const filePath = examplePath ? examplePath.replace(/^\//, '') : file.join('/')

  if (
    // Support repos whose entire purpose is to be a Next.js example, e.g.
    // https://github.com/:username/:my-cool-nextjs-example-repo-name.
    t === undefined ||
    // Support GitHub URL that ends with a trailing slash, e.g.
    // https://github.com/:username/:my-cool-nextjs-example-repo-name/
    // In this case "t" will be an empty string while the next part "_branch" will be undefined
    (t === '' && _branch === undefined)
  ) {
    try {
      const infoResponse = await fetch(
        `https://api.github.com/repos/${username}/${name}`
      )
      if (infoResponse.status !== 200) {
        return
      }

      const info = await infoResponse.json()
      return { username, name, branch: info['default_branch'], filePath }
    } catch {
      return
    }
  }

  // If examplePath is available, the branch name takes the entire path
  const branch = examplePath
    ? `${_branch}/${file.join('/')}`.replace(new RegExp(`/${filePath}|/$`), '')
    : _branch

  if (username && name && branch && t === 'tree') {
    return { username, name, branch, filePath }
  }
}

export function hasRepo({
  username,
  name,
  branch,
  filePath,
}: RepoInfo): Promise<boolean> {
  const contentsUrl = `https://api.github.com/repos/${username}/${name}/contents`
  const packagePath = `${filePath ? `/${filePath}` : ''}/package.json`

  return isUrlOk(contentsUrl + packagePath + `?ref=${branch}`)
}

export function existsInRepo(nameOrUrl: string): Promise<boolean> {
  try {
    const url = new URL(nameOrUrl)
    return isUrlOk(url.href)
  } catch {
    return isUrlOk(
      `https://api.github.com/repos/vercel/next.js/contents/examples/${encodeURIComponent(
        nameOrUrl
      )}`
    )
  }
}

async function downloadTar(url: string) {
  const tempFile = join(tmpdir(), `next.js-cna-example.temp-${Date.now()}`)

  const resp = await fetch(url)

  if (!resp.body) {
    throw new Error(`Failed to download: ${url}`)
  }

  await pipeline(
    Readable.fromWeb(resp.body as import('stream/web').ReadableStream),
    createWriteStream(tempFile)
  )

  return tempFile
}

export async function downloadAndExtractRepo(
  root: string,
  { username, name, branch, filePath }: RepoInfo
) {
  const tempFile = await downloadTar(
    `https://codeload.github.com/${username}/${name}/tar.gz/${branch}`
  )

  await tar.x({
    file: tempFile,
    cwd: root,
    strip: filePath ? filePath.split('/').length + 1 : 1,
    filter: (p) =>
      p.startsWith(
        `${name}-${branch.replace(/\//g, '-')}${
          filePath ? `/${filePath}/` : '/'
        }`
      ),
  })

  await fs.unlink(tempFile)
}

export async function downloadAndExtractExample(root: string, name: string) {
  if (name === '__internal-testing-retry') {
    throw new Error('This is an internal example for testing the CLI.')
  }

  const tempFile = await downloadTar(
    'https://codeload.github.com/vercel/next.js/tar.gz/canary'
  )

  await tar.x({
    file: tempFile,
    cwd: root,
    strip: 2 + name.split('/').length,
    filter: (p) => p.includes(`next.js-canary/examples/${name}/`),
  })

  await fs.unlink(tempFile)
}
