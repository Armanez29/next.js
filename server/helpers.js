import { join } from 'path'
import fs from 'mz/fs'
import glob from 'glob-promise'

export async function getParamRoutes (base) {
  let files = await glob(join(base, '**'))

  const result = await Promise.all(files.map(f => isFile(f)))
  files = result.reduce((_files, isFile, i) => {
    return isFile ? [..._files, files[i]] : _files
  })

  files = files.filter(f => f.indexOf('}') > 0)
  files = files.map(f => f.replace(base, ''))
  return files
}

export async function isFile (p) {
  let stat
  try {
    stat = await fs.stat(p)
  } catch (err) {
    if (err.code === 'ENOENT') return false
    throw err
  }
  return stat.isFile() || stat.isFIFO()
}

export function toParts (route) {
  return route.replace(/^\/|\.js$/g, '').split('/')
}

export function compareRoute (parts, segments) {
  // parts:    ['user', '{id}', 'action']
  // segments: ['user', '1234', 'action']
  if (parts.length === segments.length) {
    return parts.map((p, i) => p.startsWith('{') || segments[i] === parts[i]).every(ok => ok)
  }
  return false
}
