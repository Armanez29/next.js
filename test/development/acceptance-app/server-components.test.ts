/* eslint-env jest */
import { sandbox } from './helpers'
import { createNext, FileRef } from 'e2e-utils'
import { NextInstance } from 'test/lib/next-modes/base'
import path from 'path'

// TODO-APP: improve these errors
describe('Error Overlay for server components', () => {
  if (process.env.NEXT_TEST_REACT_VERSION === '^17') {
    it('should skip for react v17', () => {})
    return
  }

  let next: NextInstance

  beforeAll(async () => {
    next = await createNext({
      files: new FileRef(path.join(__dirname, 'fixtures', 'default-template')),
      dependencies: {
        react: 'latest',
        'react-dom': 'latest',
      },
      skipStart: true,
    })
  })
  afterAll(() => next.destroy())

  test('should show error when React.createContext is called', async () => {
    const { session, browser, cleanup } = await sandbox(
      next,
      new Map([
        [
          'node_modules/my-package/index.js',
          `
          const react = require('react')
          module.exports = react.createContext()
        `,
        ],
        [
          'node_modules/my-package/package.json',
          `
          {
            "name": "my-package",
            "version": "0.0.1"
          }
        `,
        ],
        [
          'app/page.js',
          `
        import Context from 'my-package'
        export default function Page() {
          return (
            <>
                <Context.Provider value="hello">
                    <h1>Page</h1>
                </Context.Provider>
            </>
          )
        }`,
        ],
      ])
    )

    // TODO-APP: currently requires a full reload because moving from a client component to a server component isn't causing a Fast Refresh yet.
    await browser.refresh()

    expect(await session.hasRedbox(true)).toBe(true)
    expect(await session.getRedboxSource(true)).toMatchSnapshot()

    await cleanup()
  })

  test('should show error when class component is called', async () => {
    const { session, browser, cleanup } = await sandbox(
      next,
      new Map([
        [
          'app/class-component.js',
          `
          import React from 'react'
          export default class Comp extends React.Component {
            render() {
                return <h1>Class Component</h1>
            }
          }
        `,
        ],
        [
          'app/page.js',
          `
        import Comp from './class-component'
        export default function Page() {
          return (
            <>
                <Comp />
            </>
          )
        }`,
        ],
      ])
    )

    // TODO-APP: currently requires a full reload because moving from a client component to a server component isn't causing a Fast Refresh yet.
    await browser.refresh()

    expect(await session.hasRedbox(true)).toBe(true)
    expect(await session.getRedboxSource(true)).toMatchSnapshot()

    await cleanup()
  })

  test('should show error when class component is called from npm package', async () => {
    const { session, browser, cleanup } = await sandbox(
      next,
      new Map([
        [
          'node_modules/my-package/index.js',
          /*
          Code below is the compiled output from Babel for the following code:
          import React from 'react'
          export default class Comp extends React.Component {
            render() {
                return <h1>Class Component</h1>
            }
          }
          */
          `
          "use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
var Comp = /*#__PURE__*/function (_React$Component) {
  _inherits(Comp, _React$Component);
  var _super = _createSuper(Comp);
  function Comp() {
    _classCallCheck(this, Comp);
    return _super.apply(this, arguments);
  }
  _createClass(Comp, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react.default.createElement("h1", null, "Hello World!");
    }
  }]);
  return Comp;
}(_react.default.Component);
exports.default = Comp;
        `,
        ],
        [
          'node_modules/my-package/package.json',
          `
          {
            "name": "my-package",
            "version": "0.0.1"
          }
        `,
        ],
        [
          'app/page.js',
          `
        import Comp from 'my-package'
        export default function Page() {
          return (
            <>
                <Comp />
            </>
          )
        }`,
        ],
      ])
    )

    // TODO-APP: currently requires a full reload because moving from a client component to a server component isn't causing a Fast Refresh yet.
    await browser.refresh()

    expect(await session.hasRedbox(true)).toBe(true)
    expect(await session.getRedboxSource(true)).toMatchSnapshot()

    await cleanup()
  })
})
