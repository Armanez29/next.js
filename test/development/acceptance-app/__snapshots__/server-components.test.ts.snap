// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Overlay for server components should show error when React.createContext is called 1`] = `
" 1 of 1 unhandled error
Server Error

TypeError: react.createContext is not a function

This error happened while generating the page. Any console logs will be displayed in the terminal window.

null"
`;

exports[`Error Overlay for server components should show error when class component is called 1`] = `
" 1 of 1 unhandled error
Server Error

TypeError: Class extends value undefined is not a constructor or null

This error happened while generating the page. Any console logs will be displayed in the terminal window.

app/class-component.js (3:44) @ React

  1 | 
  2 | import React from 'react'
> 3 | export default class Comp extends React.Component {
    |                                  ^
  4 |   render() {
  5 |       return <h1>Class Component</h1>
  6 |   }"
`;

exports[`Error Overlay for server components should show error when class component is called from npm package 1`] = `
" 1 of 1 unhandled error
Server Error

TypeError: Super expression must either be null or a function

This error happened while generating the page. Any console logs will be displayed in the terminal window.

null"
`;
